accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+All classes to be located in STJ.dll.
+
+```cs
+namespace System.Text.Json
+{
+    public partial class JsonSerializerOptions
+    {
+        // The root method that if not called, ILLinker will trim out the reference to `System.Linq.Expressions.dll`.
+        // STJ does not expose any types from SLE publically.
+        // Any System.Object-qualified type specified in a called to JsonSerializer.Deserialize() will create either a DynamicJsonArray, DynamicJsonObject or DynamicJsonValue.
+        public void EnableDynamicTypes();
+    }
+}
+```
+
+Sample code to enable dynamic types:
+```cs
+    var options = new JsonSerializerOptions();
+    options.EnableDynamicTypes();
+
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>("{\"MyProp\":42}", options);
+    int i = obj.MyProp;
+    Debug.Assert(i == 42);
+
+    string json = obj.Serialize();
+```
+
+### Varying the `T` in `Deserialize<T>`
+This
+```cs
+    dynamic obj = JsonSerializer.Deserialize<object>("{\"MyProp\":42}", options);
+```
+is equivalent to
+```cs
+    dynamic obj = JsonSerializer.Deserialize<dynamic>("{\"MyProp\":42}", options);
+```
+and these (assuming `options.EnableDynamicTypes()` was called)
+```cs
+    dynamic obj = JsonSerializer.Deserialize<DynamicJsonNode>("{\"MyProp\":42}", options);
+    dynamic obj = JsonSerializer.Deserialize<DynamicJsonObject>("{\"MyProp\":42}", options);
+```
+
+### Supporting non-dynamic to dynamic
+If a previously created non-dynamic `JsonNode` object needs dynamic support:
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    dynamic obj = node.ToDynamic();
+```
+
+The implementation of `ToDynamic()` is efficient since it will perform a shallow clone by copying the internal value (for `JsonValue`), array (for `JsonArray`) or dictionary (for `JsonObject`):
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    byte[] utf8 = node.SerializeToUtf8Bytes();
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>(utf8, options);
+```
+
+# Writable DOM
+A deserialized DOM is based upon JSON token types (object, array, number, string, true, false) since the corresponding CLR types are not known, unlike deserializing against a strongly-typed POCO. This is consistent with `JsonElement` semantics as well.
+
+Although a deserializated DOM in "read mode" fully maps to the JSON token types, once in "edit mode", CLR values to be assigned to nodes which do not necessarily map to JSON token types. Two common examples:
+- Quoted numbers. A CLR number (`Double`, `Int32`, etc) is normally serialized to JSON as a number. However, a number may be serialized as a JSON string instead - which is necessary to support CLR numbers that can't be represented in JSON as a number, such as `Double.NaN` or `Double.PositiveInfinity`.
+- String Enums. A CLR `Enum` is normally serialized to JSON as a number, but through a custom converter (including `JsonStringEnumConverter`), an `Enum` can be serialized as the string representation.
```

Comment:
I am surprised to see support for these as first-class features within the dom. Do we have evidence that folks really want this, outside of the serializer? I consider the DOM to be the true representation of JSON and hence questions around which CLR type does each token correspond to isn't really fundamental to the object hierarchy. Converting to CLR types is more of the responsibility of the serializer and the helper functions on JsonNode/JsonElement, rather than using the converter approach.

These are more edge case uses of the DOM. We should prioritize the mainline scenario of creating, modifying, and querying the DOM over these features.

Reply:
It comes because we allow any CLR value (such as an Enum) to be specified in `JsonValue<T>`. Thus we need to address the mapping from CLR types to JSON.

Reply:
If we tighten the API to not allow any value, and thus have a predictable token type, at a minimum we likely need to add some more built-in types like Enum. For example, this works in Newtonsoft:
```cs
JToken j = new JValue(ConsoleColor.Red);
string json = j.ToString();
Debug.Assert(json == "Red");
```

Reply:
> We should prioritize the mainline scenario of creating, modifying, and querying the DOM over these features.

What DOM querying features are you thinking or referring to (other than IEnumerable support on JsonArray\JsonObject)?